cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+//! Overlap detection for rules in ISLE.
+
+use crate::error::{Error, Result, Source, Span};
+use crate::sema::{
+    self, ConstructorKind, Rule, RuleId, Sym, Term, TermEnv, TermId, TermKind, Type, TypeEnv,
+    TypeId, VarId,
+};
+
+/// Check for overlap.
+pub fn check(tyenv: &TypeEnv, termenv: &TermEnv) -> Result<()> {
+    let env = Env::new(tyenv, termenv);
+    let mut errors = Errors::new(termenv.rules.len());
+    for term in termenv.terms.iter() {
+        // The only isle declaration that currently produces overlap is constructors whose
+        // definition is entirely in isle.
+        if !env.is_internal_constructor(term.id) {
+            continue;
+        }
+
+        check_overlap_groups(&mut errors, &env, term);
+    }
+
+    if !errors.is_empty() {
+        let mut errors = errors.report(&env);
+        return match errors.len() {
+            1 => Err(errors.pop().unwrap()),
+            _ => Err(Error::Errors(std::mem::take(&mut errors))),
+        };
+    }
+
+    Ok(())
+}
+
+/// A node in the error graph.
+struct Node {
+    /// The number of other rules this node overlaps with.
+    degree: usize,
+
+    /// `true` entries where an edge exists to the node at that index.
+    edges: Vec<bool>,
```

Comment:
I used a HashSet to begin with, and it was considerably slower when building without `--release`. The switch to this dense representation made the runtime go from 55s for the x64 lowering down to about 10s. One thought I had was that the vector could be an `Option<Vec<Bool>>` where it would only allocate the bool vector if there was overlap present involving that rule. Another thought I had was that something like c++'s specialization of `std::vector<bool>` would be really nice here.

Reply:
The motivation for improving performance when building without `--release` is that cranelift currently builds the isle source without `--release` in the cranelift-codegen build.rs.

Reply:
You don't need to wrap the `Vec` in an `Option`; `Vec::new()` already doesn't heap-allocate, so you can just use `Vec::is_empty` as your flag instead of `Some`/`None`. And yeah, I keep wanting to use the `bitvec` crate for bit-packed boolean vectors, but we don't use it anywhere else and we don't add dependencies if we can help it.

But also: wat? How could HashSet be that much slower? I feel like there must have been something else going on. (While trying to come up with hypotheses that might explain the difference, I worked out that the HashSet is smaller than the `Vec<bool>` as long as a rule overlaps with less than 11% of the other rules, but that doesn't seem likely to answer the question.)

Also, in the long run I don't want to optimize for the case where there are a bunch of overlap errors if doing that makes things worse when there aren't any errors. While we're fixing the current backlog of overlaps, of course, it's reasonable to try to keep the debug loop short. But on the other hand, we can use `--release` builds of islec for that, right?

Reply:
My assumption about why the vector is faster is that all the allocation is done up front. I don't know the specifics of how `HashSet` is implemented under the hood, but a lot of insertions to different rows could cause an allocation for each insertion, right?

I agree that it's not worth over-optimizing for the case where we have lots of overlap, but we're talking about 1MB of memory to track the overlaps in the x64 lowerings. Paired with the one-off allocation strategy when we need to start adding overlap edges, it feels like this is a reasonable strategy to keep, but if you'd rather go back to the `HashSet` I'm not opposed.

Reply:
I reworked the overlap checking to work over pairs of related rules instead, which is much faster and reported more accurate overlap information.